llumnix/backends/vllm/migration_backend.py

```diff
@@ -0,0 +1,264 @@
+            for layer_idx in range(self.num_layers):
+                self.cache_engine.attn_backend.swap_blocks(self.gpu_cache[layer_idx], send_cache[layer_idx], src_to_dst)
+        torch.cuda.Stream.synchronize(self.migration_stream)
+        return send_cache.to(self.rpc_dtype).numpy()
+
+    def do_recv(self, src_handle, blocks: List[int]):
+        num_blocks = len(blocks)
+        src_to_dst = dict(enumerate(blocks))
+        recv_cache = self.dummy_cache[:num_blocks].view(self.num_layers, 2, num_blocks, self.migration_cache_size)
+        # use pin memory dummy_cache to speed up data transfer
+        recv_cache.copy_(torch.from_numpy(src_handle))
+
+        with torch.cuda.stream(self.migration_stream):
+            for layer_idx in range(self.num_layers):
+                self.cache_engine.attn_backend.swap_blocks(recv_cache[layer_idx], self.gpu_cache[layer_idx], src_to_dst)
+        torch.cuda.Stream.synchronize(self.migration_stream)
+
+def try_import_gloo():
+    try:
+        # pylint: disable=C0415
+        from ray.util.collective.collective_group import gloo_util
+        import pygloo
+
+        # Add support for bf16 type in Gloo. Now bf16 and fp16 both map to glooFloat16, but this is okay because
+        # Gloo only uses the data type size for transmission.
+        # pylint: disable=W0221,I1101
+        gloo_util.TORCH_GLOO_DTYPE_MAP[torch.bfloat16] = pygloo.glooDataType_t.glooFloat16
+    except ImportError as e:
+        raise ImportError("Gloo is not installed. Please install it first.") from e
+
+class RayColMigrationBackend(MigrationBackendBase):
+    def __init__(self, migration_config: MigrationConfig, cache_engine: CacheEngine, local_rank,
+                 scheduling_strategy, is_driver_worker, gpu_cache) -> None:
+        super().__init__()
+
+        self.migration_config = migration_config
+        self.cache_engine = cache_engine
+        self.backend = migration_config.migration_backend
+        self.migration_num_layers = min(migration_config.migration_num_layers, self.cache_engine.num_layers)
+        self.num_migration_cache_blocks = migration_config.migration_cache_blocks
+
+        self.backend = migration_config.migration_backend
+        self.global_world_size = -1
+        self.global_rank = -1
+        self.group_name = None
+
+        self.local_rank = local_rank
+        self.actor = ProxyActor.options(scheduling_strategy=scheduling_strategy).remote()
+        self.is_driver_worker = is_driver_worker
+        self.gpu_cache = gpu_cache
+
+        self.migration_cache_size = self.cache_engine.block_size * self.cache_engine.num_heads * self.cache_engine.head_size
+
+        if self.backend == 'gloo':
+            try_import_gloo()
+            self.cache_device = "cpu"
+        elif self.backend == 'nccl':
+            self.cache_device = torch.device(f"cuda:{self.local_rank}")
+        else:
+            raise ValueError("backend must be 'gloo' or 'nccl'")
```

Comment:
This is unnessary, because we only initialize it when backend is gloo or nccl.

Reply:
get